lightningd/log.c

```diff
@@ -564,40 +564,66 @@ void logv(struct logger *log, enum log_level level,
 	  const char *fmt, va_list ap)
 {
 	int save_errno = errno;
-	struct log_entry *l = new_log_entry(log, level, node_id);
 	char *log_msg = NULL;
 
 	/* This is WARN_UNUSED_RESULT, because everyone should somehow deal
 	 * with OOM, even though nobody does. */
 	if (vasprintf(&log_msg, fmt, ap) == -1)
 		abort();
 
-	char **lines = tal_strsplit(
-	    log, json_escape_unescape(log, (struct json_escape *)log_msg), "\n",
-	    STR_NO_EMPTY);
+	const char *unescaped_log =
+	    json_escape_unescape(NULL, (struct json_escape *)log_msg);
+	if (!strchr(log_msg, '\\') || !unescaped_log) {
+		struct log_entry *l = new_log_entry(log, level, node_id);
 
-	/* Split to lines and log them separately. */
-	for (size_t j = 0; lines[j]; j++) {
-		l->log = tal_strdup(log, lines[j]);
+		l->log = strdup(log_msg);
+		size_t log_len = strlen(l->log);
 
 		/* Sanitize any non-printable characters, and replace with '?'
 		 */
-		size_t line_len = strlen(l->log);
-		for (size_t i = 0; i < line_len; i++)
+		for (size_t i = 0; i < log_len; i++)
 			if (l->log[i] < ' ' || l->log[i] >= 0x7f)
 				l->log[i] = '?';
 
 		maybe_print(log, l);
 		maybe_notify_log(log, l);
 
 		add_entry(log, &l);
-	}
 
-	if (call_notifier)
```

Comment:
Extract this common code into a new function, rather than calling twice.  Say "add_one_log_entry(log, level, nodeid, string)", which does new_log_entry, strdup, sanitize, maybe_print, maybe_notify_log, add_entry.

Then you can make the code a bit more optimal, and avoid json_escape if there are no escapes:

```
/* Nothing to escape: simple copy */
if (!strchr(log_msg, '\\'))
   add_one_log_entry(log, level, nodeid, log_msg);
/* If it's weird, unescaping can fail */
else if ((unescaped_log = json_escape_unescape(tmpctx, (struct json_escape *)log_msg)) == NULL) 
   add_one_log_entry(log, level, nodeid, log_msg);
else {
  char **lines = tal_strsplit(unescaped_log, unescaped_log, "\n", STR_EMPTY_OK);
	/* Split to lines and log them separately. */
		for (size_t i = 0; lines[i]; i++)
                     add_one_log_entry(log, level, nodeid, lines[i]);
}